sysl2/sysl/importer/grammar.go

```diff
@@ -0,0 +1,205 @@
+	case *ebnfGrammar.Quantifier_ZeroPlus:
+		fieldType = fmt.Sprintf("sequence of %s?", fieldType)
+	case *ebnfGrammar.Quantifier_OnePlus:
+		fieldName = fmt.Sprintf("%s(1..)", fieldName)
+	default:
+		g.logger.Warnf("unexpected quantifier type")
+	}
+	return Field{
+		Name: fieldName,
+		Type: &StandardType{
+			name: fieldType,
+		}}
+}
+
+func (g *Grammar) createRule(rule *ebnfGrammar.Rule, name string) Type {
+	ruleType := StandardType{
+		name: name,
+	}
+	if !g.visitedMap[name] {
+		g.visitedMap[name] = true
+	}
+	for _, seq := range rule.Choices.Sequence {
+		for _, term := range seq.Term {
+			switch x := term.Atom.Union.(type) {
+			case *ebnfGrammar.Atom_String_:
+				// hit the leaf node
+				// return
+			case *ebnfGrammar.Atom_Choices:
+				choiceField := g.quantifyUnionType("__Choice_"+name, "__Choice_"+name, term.Quantifier)
+				ruleType.Properties = append(ruleType.Properties, choiceField)
+
+				// visit the unvisited node
+				if !g.visitedMap["__Choice_"+name] {
+					childChoiceType := g.createChoice(x.Choices, "__Choice_"+name)
+					g.types.types = append([]Type{childChoiceType}, g.types.types...)
+				}
+			case *ebnfGrammar.Atom_Rulename:
+				var childRuleType Type
+				typeString := "string"
+				if _, exists := g.grammar.Rules[x.Rulename.GetName()]; exists {
+					typeString = x.Rulename.Name
+				}
+				ruleField := g.quantifyUnionType(x.Rulename.Name, typeString, term.Quantifier)
+				ruleType.Properties = append(ruleType.Properties, ruleField)
+
+				// visit the unvisited node
+				if !g.visitedMap[x.Rulename.GetName()] {
+					if rule, exists := g.grammar.Rules[x.Rulename.GetName()]; exists {
+						childRuleType = g.createRule(g.grammar.Rules[rule.Name.Name], rule.Name.Name)
+						g.types.types = append([]Type{childRuleType}, g.types.types...)
+					}
+				}
+			default:
+				g.logger.Warnf("unexpected term type")
+			}
+		}
+	}
+	var baseType Type = &ruleType
+	return baseType
+}
+
+func (g *Grammar) writeSysl() (string, error) {
+	g.grammar = parser.ParseEBNF(g.grammarParam.grammarInput, g.grammarParam.grammarName,
+		g.grammarParam.startRule)
+	rule := g.grammar.Rules[g.grammarParam.startRule]
+	if rule == nil {
+		return "", fmt.Errorf("start rule not found")
+	}
+	var outputData OutputData
+	var syslBytes bytes.Buffer
+	syslWriter := newWriter(&syslBytes, g.logger)
+	switch g.grammarParam.startRule {
+	case `goFile`:
+		outputData = OutputData{
+			AppName: "Go",
+		}
+	case `javaFile`:
+		outputData = OutputData{
+			AppName: "Java",
+		}
+	default:
+		return "", fmt.Errorf("unexpected entry rule")
+	}
+	info := SyslInfo{
+		OutputData:  outputData,
+		Title:       g.grammarParam.startRule,
+		Description: outputData.AppName + " Grammar",
+	}
+
+	startRule := g.grammar.GetRules()[g.grammarParam.startRule]
+	ruleType := g.createRule(startRule, g.grammarParam.startRule)
+	g.types.types = append([]Type{ruleType}, g.types.types...)
+	g.visitedMap[g.grammarParam.startRule] = true
```

Comment:
I dont understand the need to specify the start rule,  The sysl file should contain every rule found in the grammar, not just the ones accessable from the start rule.

Reply:
Well according to my understanding the AST for the chosen grammar would have only 1 root, and that is the start rule. We could deduce the start rule automatically rather than the user passing it over the command line.

Reply:
Anyways the intent is to use it more from codegen rather than importer.

Reply:
It wanting to be consumed by the codeegen code, and being able to geneerate sysl from grammar are two related but not the same thing. The sysl generation doesnt need to know the start rule.

Reply:
Taken care